src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java

```diff
@@ -0,0 +1,191 @@
+import org.apache.hc.client5.http.fluent.Response;
+import org.apache.hc.core5.http.ContentType;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.common.returnsreceiver.qual.This;
+import org.jetbrains.annotations.TestOnly;
+
+public class OpenAIPlugin<T extends Message> implements LLMPlugin<T> {
+
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+  private static final String ENDPOINT = "https://api.openai.com/v1/chat/completions";
+  private final OpenAIConfig config;
+  private final Encoding tokenEncoding;
+  private final int tokensPerMessage;
+  private final int tokensPerName;
+  private URI endpoint;
+
+  public OpenAIPlugin(OpenAIConfig config) {
+    this.config = config;
+
+    try {
+      this.endpoint = new URI(ENDPOINT);
+    } catch (URISyntaxException e) {
+      throw new RuntimeException(e); // this should be impossible
+    }
+    tokenEncoding =
+        Encodings.newDefaultEncodingRegistry()
+            .getEncodingForModel(config.model().properties().jtokkinModel());
+
+    switch (config.model()) {
+      case GPT4, GPT432K -> {
+        tokensPerMessage = 3;
+        tokensPerName = 1;
+      }
+      case GPT35TURBO, GPT35TURBO16K -> {
+        tokensPerMessage = 4; // every message follows <|start|>{role/name}\n{content}<|end|>\n
+        tokensPerName = -1; // if there's a name, the role is omitted
+      }
+      default -> throw new IllegalArgumentException("Unsupported model: " + config.model());
+    }
+  }
+
+  @TestOnly
+  public @This OpenAIPlugin<T> endpoint(URI endpoint) {
+    this.endpoint = endpoint;
+    return this;
+  }
+
+  private int tokenCount(JsonNode message) {
+    int tokenCount = tokensPerMessage;
+    tokenCount += tokenEncoding.countTokens(message.get("content").textValue());
+    tokenCount += tokenEncoding.countTokens(message.get("role").textValue());
+    @Nullable JsonNode name = message.get("name");
+    if (name != null) {
+      tokenCount += tokenEncoding.countTokens(name.textValue());
+      tokenCount += tokensPerName;
+    }
+    return tokenCount;
+  }
+
+  private Optional<ArrayNode> pruneMessages(ArrayNode messages, @Nullable JsonNode functions)
+      throws JsonProcessingException {
+
+    int functionTokens = 0;
+    if (functions != null) {
+      // This is honestly a guess, it's undocumented
+      functionTokens = tokenEncoding.countTokens(MAPPER.writeValueAsString(functions));
+    }
+
+    ArrayNode output = MAPPER.createArrayNode();
+    int totalTokens = functionTokens;
+    totalTokens += 3; // every reply is primed with <|start|>assistant<|message|>
+
+    JsonNode systemMessage = messages.get(0);
+    boolean hasSystemMessage = systemMessage.get("role").textValue().equals("system");
+    if (hasSystemMessage) {
+      // if the system message is present it's required
+      totalTokens += tokenCount(messages.get(0));
+    }
+    for (int i = messages.size() - 1; i >= 0; i--) {
```

Comment:
Not necessarily cleaner but we can remove the above if statement on 101 and the continue on 109 if its for (int i = hasSystemMessage ? 0 : int i = messages.size() - 1; i >= hasSystemMessage ? -messages.size() + 2 : 0; i--)

Reply:
Would also need to change the input on 115/118 though so maybe not worth it

Reply:
yea it's definitely not an elegant way to code this, but I had trouble find an elegant way to do it